### PR TITLE
Swap interface in Telegram bot

### PR DESCRIPTION
**Pull request by antonyshchenko**
- bot is now modelled as a finite state machine
- pending airdrops, deposits, withdrawals and orders are tracked by a separate thread w/o relying on websockets

Created separate tickets for:
- market price should be calculated based on the order book and amount, similar to how it is done in web UI
- should consider limits and fees before allowing to submit an order

https://github.com/Chainring-Inc/chainring-monorepo/assets/443906/c76b0742-01d6-4aff-87be-6dd5c650e559

### CODE REVIEW

**Comment by bholzman** [comment]
> This is more of a draft, to get some feedback on the interface. There are some pieces missing:
> 
> * market price should be calculated based on the order book and amount, similar to how it is done in web UI
> * should consider limits and fees before allowing to submit an order
> 
>  tg-swap-s.mov

So I think we should completely remove the concept of "Market" from here. What if it was like this -- there's just a button "Swap" which when you click it there's a list of assets to sell; when they click one, there's a list of assets to buy, and then they just have to give the amount to do the swap. Then, we could remember their "sell token" and "buy token" and give them a "Swap more BTC for ETH" (or whatever) button.

**Comment by antonyshchenko** [comment]
I like that @bholzman. I will give it a try tomorrow. Here is an example how I see the flow:

```
1. User enters /start command
2. Bot shows main menu with wallet address, balances and buttons: [Swap] [Deposit] [Withdraw] [Settings]
3. User taps [Swap]
4. Bot: Select token to sell: [BTC] [ETH] ... [USDC] [Cancel]
5. User taps [BTC]
6. Bot: Select token to sell: [ETH] ... [USDC] [Cancel]
7. User taps [ETH] 
8. Bot: You have X BTC. How much are you willing to swap?
9. User enters amount X
10. Bot: Selling X BTC will give you Y ETH at the price of Z ETH with a fee of Q ETH. Proceed? [Confirm] [Cancel] [Change amount]
11. User taps [Confirm]
12. Bot: Order submitted
13. Bot: Order filled
14. Bot: Trade settled
15. Bot: shows main menu with balances and buttons: [Swap more BTC for ETH] [Swap ETH for BTC] [Swap other tokens] [Deposit] [Withdraw] ...
```

I'm not sure how step 8 and 10 should look like when user is willing to sell quote symbol, say ETH in the BTC/ETH market. We have to consider that fee is paid in quote currency, so we can't just say `You have X ETH. How much are you willing to swap?` Or perhaps we can, but should reject the input with an explanation if what they have entered + fee is more than what they have. Any thoughts on this?

**Comment by bholzman** [comment]
> I like that @bholzman. I will give it a try tomorrow. Here is an example how I see the flow:
> 
> ```
> 1. User enters /start command
> 2. Bot shows main menu with wallet address, balances and buttons: [Swap] [Deposit] [Withdraw] [Settings]
> 3. User taps [Swap]
> 4. Bot: Select token to sell: [BTC] [ETH] ... [USDC] [Cancel]
> 5. User taps [BTC]
> 6. Bot: Select token to sell: [ETH] ... [USDC] [Cancel]
> 7. User taps [ETH] 
> 8. Bot: You have X BTC. How much are you willing to swap?
> 9. User enters amount X
> 10. Bot: Selling X BTC will give you Y ETH at the price of Z ETH with a fee of Q ETH. Proceed? [Confirm] [Cancel] [Change amount]
> 11. User taps [Confirm]
> 12. Bot: Order submitted
> 13. Bot: Order filled
> 14. Bot: Trade settled
> 15. Bot: shows main menu with balances and buttons: [Swap more BTC for ETH] [Swap ETH for BTC] [Swap other tokens] [Deposit] [Withdraw] ...
> ```
> 
> I'm not sure how step 8 and 10 should look like when user is willing to sell quote symbol, say ETH in the BTC/ETH market. We have to consider that fee is paid in quote currency, so we can't just say `You have X ETH. How much are you willing to swap?` Or perhaps we can, but should reject the input with an explanation if what they have entered + fee is more than what they have. Any thoughts on this?

I think probably a "Sell Max" button would handle it most elegantly. Obviously wait until we've implemented that in the sequencer to do it here. We could show an error as you suggest, I think that makes sense.

File: backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        val messageIdsForDeletion = array<Int>("message_ids_for_deletion").default(emptyList())
+        val sessionState = jsonb<V43_BotSessionState>("session_state", KotlinxSerialization.json).default(V43_BotSessionState.Initial)
+    }
+
+    override fun run() {
+        transaction {
+            exec("ALTER TABLE telegram_bot_user ADD COLUMN updated_at timestamp without time zone")
+            exec("UPDATE telegram_bot_user SET updated_at = now()")
+            exec("ALTER TABLE telegram_bot_user ALTER COLUMN updated_at SET NOT NULL")
+            SchemaUtils.createMissingTablesAndColumns(V43_TelegramBotUserTable)
```
Comment (bholzman) [inline backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt:41]: fwiw couldn't this be first and then we wouldn't need the "ADD COLUMN updated_at" line?

**Review by bholzman — APPROVED** [review]
Looks great! FSM model makes it very easy to follow.

File: backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt (hunk 1 above)
Comment (antonyshchenko) [reply backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt:41]: The column is non nullable, so it has to be done in 3 steps:
- add a nullable column
- set value for existing records
- change it to non nullable

Comment (bholzman) [reply backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt:41]: Yes I know, but you defined it as nullable in the table here so you could use schemautils to create it and then the two commands to update and make it not nullable

Comment (antonyshchenko) [reply backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt:41]: Here is how it is defined in the migration:
```
 val updatedAt = timestamp("updated_at")
```

Defining it as nullable would look like this:
```
 val updatedAt = timestamp("updated_at").nullable()
```

Comment (bholzman) [reply backend/src/main/kotlin/co/chainring/core/model/db/migrations/V43_AddSessionStateToTelegramBotUser.kt]: True, true. I guess I’ve usually done it that way, but it doesn’t really make a difference. Nevermind.

**Pull request merged**